src/librustc/hir/lowering.rs

```diff
@@ -5495,3 +5495,65 @@ fn body_ids(bodies: &BTreeMap<hir::BodyId, hir::Body>) -> Vec<hir::BodyId> {
     body_ids.sort_by_key(|b| bodies[b].value.span);
     body_ids
 }
+
+/// This function checks if the specified expression is a built-in range literal.
+/// (See: `LoweringContext::lower_expr()`).
+pub fn is_range_literal(sess: &Session, expr: &hir::Expr) -> bool {
+    use hir::{Path, QPath, ExprKind, TyKind};
+
+    // We support `::std::ops::Range` and `::core::ops::Range` prefixes.
+    let is_range_path = |path: &Path| {
+        let mut segs = path.segments.iter().map(|seg| seg.ident.as_str());
+
+        if let (Some(root), Some(std_core), Some(ops), Some(range), None) =
+            (segs.next(), segs.next(), segs.next(), segs.next(), segs.next())
```

Comment:
😭 

Collecting into an array would have been so nice...

Maybe leave a FIXME for when we can do so?

Reply:
This entire section could also be rewritten as

```
        let path = (segs.next(), segs.next(), segs.next(), segs.next(), segs.next());
        if let (Some("{{root}}"),  Some(std_core), Some("ops"), Some(range)) = path {
            std_core == "std" || std_core == "core" && range.starts_with("Range")
        } else {
            false
        }
```

Reply:
You can use `.collect::<Vec<_>>()` (or even some sort of `ArrayVec`) and then pattern-match it with `["{{root}}", std_core, "ops", range]`.

But also, doesn't clippy have something for this? cc @oli-obk